go/libkb/leveldb.go

```diff
@@ -11,56 +11,145 @@ import (
+	batch := new(leveldb.Batch)
+	idb := id.ToBytes(levelDbTableKv)
+	batch.Put(idb, value)
+	if aliases != nil {
+		for _, alias := range aliases {
+			batch.Put(alias.ToBytes(levelDbTableLo), idb)
+		}
+	}
+
+	return ops.Write(batch, nil)
+}
+
+func levelDbGetWhich(ops levelDBOps, id DbKey, which string) (val []byte, found bool, err error) {
+	val, err = ops.Get(id.ToBytes(which), nil)
+	found = false
+	if err == nil {
+		found = true
+	} else if err == leveldb.ErrNotFound {
+		err = nil
+	}
+	return val, found, err
+}
+
+func levelDbGet(ops levelDBOps, id DbKey) (val []byte, found bool, err error) {
+	return levelDbGetWhich(ops, id, levelDbTableKv)
+}
+
+func levelDbLookup(ops levelDBOps, id DbKey) (val []byte, found bool, err error) {
+	val, found, err = levelDbGetWhich(ops, id, levelDbTableLo)
+	if found {
+		if tab, id2, err2 := DbKeyParse(string(val)); err2 != nil {
+			err = err2
+		} else if tab != levelDbTableKv {
+			err = fmt.Errorf("bad alias; expected 'kv' but got '%s'", tab)
+		} else {
+			val, found, err = levelDbGetWhich(ops, *id2, levelDbTableKv)
+		}
+	}
+	return val, found, err
+}
+
+func levelDbDelete(ops levelDBOps, id DbKey) error {
+	return ops.Delete(id.ToBytes(levelDbTableKv), nil)
+}
+
 type LevelDb struct {
-	db       *leveldb.DB
+	// We use a RWMutes here to ensure close doesn't happen in the middle of
+	// other DB operations, and DB operations doesn't happen after close. The
+	// lock should be considered for the db pointer and dbOpenerOnce pointer,
+	// rather than the DB itself.  More specifically, close does Lock(), while
+	// other DB operations does RLock().
+	sync.RWMutex
+	db           *leveldb.DB
+	dbOpenerOnce *sync.Once
+
 	filename string
-	sync.Mutex
 	Contextified
 }
 
 func NewLevelDb(g *GlobalContext, filename func() string) *LevelDb {
 	return &LevelDb{
 		Contextified: NewContextified(g),
 		filename:     filename(),
+		dbOpenerOnce: new(sync.Once),
 	}
 }
 
 // Explicit open does nothing we'll wait for a lazy open
 func (l *LevelDb) Open() error { return nil }
 
-func (l *LevelDb) open() error {
-	l.Lock()
-	defer l.Unlock()
-
-	var err error
-	if l.db == nil {
-		l.G().Log.Debug("+ LevelDb.open")
-		fn := l.GetFilename()
-		l.G().Log.Debug("| Opening LevelDB for local cache: %v %s", l, fn)
-		l.db, err = leveldb.OpenFile(fn, nil)
-		if err != nil {
-			if _, ok := err.(*errors.ErrCorrupted); ok {
-				l.G().Log.Debug("| LevelDB was corrupted; attempting recovery (%v)", err)
-				l.db, err = leveldb.RecoverFile(fn, nil)
-				if err != nil {
-					l.G().Log.Debug("| Recovery failed: %v", err)
-				} else {
-					l.G().Log.Debug("| Recovery succeeded!")
+func (l *LevelDb) doWhileOpenAndNukeIfCorrupted(action func() error) (err error) {
+	err = func() error {
+		l.RLock()
+		defer l.RUnlock()
+
+		// This only happens at first ever doWhileOpenAndNukeIfCorrupted call, or
+		// when doOpenerOnce is just reset in Nuke()
+		l.dbOpenerOnce.Do(func() {
```

Comment:
The `Once` makes it possible to use a `RWMutex`, but still allowing multiple routine to call this at the same time and ensuring only one gets to do the actual opening